[mkcal] Purge incidence from given notebook.

dcaliste opened the pull request:
Adding a notebook UID to the purgeDeletedIncidences() API would allow to selectively remove incidences from the database. Without it, it is not possible to leave
incidences marked as deleted in a notebook different than the one we want to insert an incidence to with the same UID (like when changing notebook for instance).

@pvuorela , this is not a definitive answer to the problem of switching notebook for an incidence, but it may help to. Currently, the `purgeDeletedIncidences()` API is not aware of the notebook the incidence belongs to and purge all incidences marked as deleted based on UID/recID only.

This `purgeDeletedIncidences()` API is used in various places:
* when inserting a new incidence. Indeed, we need to remove all incidences previously marked as deleted with the same UID, otherwise it may confuse the sync routines, having an inserted and a deleted same incidence (recreation via an import for instance of the same event after having deleted it, or recreation via sync after a local deletion but with upstream changes for instance)… This is done internally in mKCal on DB insert. Here, the purge should be done only for incidence of the notebook we are going to insert into. This PR is addressing this case.
* when an incidence is marked as deleted on a local notebook. Indeed, there is no need to keep the incidence in DB because nothing will sync with a local notebook and there thus is no need to list deleted incidences. This is done in nemo-qml-plugin-calendar and can be addressed by a separated PR there once the added API here has been validated.
* when a sync plugin after a local deletion has been propagated upstream. There is no need to keep the deleted incidence on device. Here also, the purge should be done for the notebook we are syncing with. It can be addressed by separated PRs in the sync plugins once the new API here has been validated.

If you agree this will end up with all user of `purgeDeletedIncidences()` to use the notebook optional argument. At that point, I can make this argument mandatory. Indeed, the use of the API with an empty notebookUid argument is very dangerous since it removes all deleted incidences irrespective of their notebook belonging. If you prefer, I can prepare the PRs there in the same time and you may accept all at once. This would simplify a bit the PR here since it would remove the branching in case the notebook UID argument is empty.

pvuorela commented:
I'm not sure if this would help enough the case of moving incidences between notebooks. Problem being that Calendar in kf5-calendarcore level  enforces that there cannot be same incidence on different notebooks. And since a deleted instance can only belong to a single notebook, it should be safe to call already?

To properly fix that, I'd suspect we'd eventually need to split each notebook into a separate calendar+storage pair, and if done so, there wouldn't be need for such notebook parameters anymore for the purge.

dcaliste commented:
> Problem being that Calendar in kf5-calendarcore level enforces that there cannot be same incidence on different notebooks. 

When changing notebook, there is still only one non-deleted incidence for a given UID at every moment. So it's not violating any constraints from KCalendarCore. At least not on the UID unicity.

> And since a deleted instance can only belong to a single notebook, it should be safe to call already?

Well, no, because this purge API is removing from every notebook. But otherwise, in mKCal, there is no constrain to have a deleted incidence to belong to a unique notebook. You can have incidence UID=123-456, deleted=20230421T1643, Notebook=ABC-DEF and at the same time an incidence UID=123-456, deleted=20230421T1644, Notebook=HG-IJK.

The idea is to allow this scheme at mKCal level :
* Have an incidence 123-456 from Notebook ABC-DEF,
* Move it to notebook HG-IJK and save,
* in mKCal then, it is adding dateDeleted = 20230421T1647 (current time) to this incidence entry, and it is creating a new incidence entry 123-456 for notebook HG-IJK.

This is not violating any rule in DB, nor in calendar. Fetching UID 123-456 from calendar will return the new incidence. The new incidence will also being seen as "created" for sync plugin that may handle notebook HG-IJK. As the deleted entry in the DB will be advertised to the sync plugin handling notebook ABC-DEF. The later sync plugin will then purge this deleted entry (thus the current PR).

The action is "reentrant". I mean, you can continue to move incidence 123-456 to yet another notebook, for instance UVW-XYZ. Or even put it back to the original ABC-DEF. In the former case, we will end up with:
- a new entry for incidence 123-456 notebook UVW-XYZ,
- a deleted entry for incidence 123-456 notebook ABC-DEF,
- a deleted entry for incidence 123-456 notebook HG-IJK.

And for the later case :
- a recreated entry for incidence 123-456 for notebook ABC-DEF,
- a deleted entry for incidence 123-456 in notebook HG-IJK.

The current PR does not allow what I'm describing here, but it would help by not purging all deleted incidences when called.

> To properly fix that, I'd suspect we'd eventually need to split each notebook into a separate calendar+storage pair,

That's appealing, but it raises a lot of issues. I started to think about it but it means that the ExtendedCalendar::load(uid) kind of call is not possible anymore. And it's used in many places from sync plugins to QML bindings. It requires to add notebookUid argument in many places and that's really a lot of changes.

Thinking more about it, it seems that the current situation is not that bad : UID for incidence should be unique throughout all notebook. It requires to adjust UIDs received from external sources like sync plugins (which is already done, so no work). But otherwise, it's very convenient because we can have a multi calendar access with only one string : the UID (in fact the instanceIdentifier). The issue of notebook change in the current scheme is not violating the UID unicity : there is only one non deleted incidence with a given UID at the same time. The problem arises with the deleted incidences : while we can have several deleted incidences (also in various notebooks), the purging of then is too large to allow to keep track of deleted incidences from other notebooks.

pvuorela commented:
We discussed this on IRC some time ago with the plan of separating the notebooks better, but not sure did we reach final conclusion regarding this PR.

So I think you're right in the sense that mkcal actually supports same event deleted in multiple calendars at the time (as long as they are not deleted at the same time, eh?). A bit suspicious about the kf5-calendarcore having methods without notebook parameters, but guess we are more or less using the mkcal variants with the notebook.

So in short term I could be ok merging this if you see it good, but there is still the bigger picture.

dcaliste commented on src/sqliteformat.cpp:
```diff
@@ -296,11 +298,12 @@ bool SqliteFormat::modifyCalendars(const Notebook &notebook,
     int rv;
     int index = 1;
     const QByteArray u(incidence.uid().toUtf8());
+    const QByteArray n(notebook.toUtf8());
```
Sure, changed for more meaningful variable names.

dcaliste commented:
> So I think you're right in the sense that mkcal actually supports same event deleted in multiple calendars at the time (as long as they are not deleted at the same time, eh?).

Indeed, the deleted date must be different at the moment. But it's just a matter of dropping the current UNIQUE INDEX and recreate it with the Notebook column in addition.

> A bit suspicious about the kf5-calendarcore having methods without notebook parameters, but guess we are more or less using the mkcal variants with the notebook.

What do you mean ? Which upstream methods are you referring to ? This purgeDeletedIncidences() was an addition of my own some years ago to avoid always cluttering the DB with incidences marked as deleted. To bad that I didn't think multi notebooks at that time...

> So in short term I could be ok merging this if you see it good, but there is still the bigger picture.

Yes, definitely, this is not a complete answer to the multi-notebook issue. It is contributing to though by modifying the API of a routine that was not multi-notebook aware and that could be harmful when used without the notebookUid argument in multi-notebook contexts. As you can see in #60 there are other minor fixes that are required here and there to be sure that incidence UIDs are not the only way to access the data but that notebook UID are also required (but fxies there actually don't change the public API).

dcaliste commented:
Thank you. I'll create minor PRs in Buteo sync plugins and nemo-qml-plugin-calendar where the `purgeDeletedIncidences()` is used without Notebook argument and thus now raising a warning.
